src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs

```diff
@@ -894,103 +925,337 @@ private bool CheckPropertyValueKind(SyntaxNode node, BoundExpression expr, BindV
                     {
-                        argument = args[paramIndex];
+                        Debug.Assert(refKind == RefKind.Ref || refKind == RefKind.None);
+                        escapeTo = Math.Min(escapeTo, GetValEscape(argument, scopeOfTheContainingExpression));
                     }
                 }
-                else
+            }
+
+            if (escapeTo == scopeOfTheContainingExpression)
+            {
+                // cannot fail. common case.
+                return true;
+            }
+
+            // check that no ref/out/in or ref-like can be captured from a narrower scope
+            ArrayBuilder<bool> inParametersMatchedWithArgs = null;
+            if (!args.IsDefault)
+            {
+                for (var argIndex = 0; argIndex < args.Length; argIndex++)
                 {
-                    for (int argIndex = 0; argIndex < args.Length; argIndex++)
+                    var effectiveRefKind = argRefKinds.IsDefault ? RefKind.None : argRefKinds[argIndex];
+                    if (effectiveRefKind == RefKind.None && argIndex < parameters.Length)
                     {
-                        if (argToParamsOpt[argIndex] == paramIndex)
+                        var paramIndex = argsToParamsOpt.IsDefault ? argIndex : argsToParamsOpt[argIndex];
+                        if (parameters[paramIndex].RefKind == RefKind.RefReadOnly)
                         {
-                            argument = args[argIndex];
-                            break;
+                            effectiveRefKind = RefKind.RefReadOnly;
+                            inParametersMatchedWithArgs = inParametersMatchedWithArgs ?? ArrayBuilder<bool>.GetInstance(parameters.Length, fillWithValue: false);
+                            inParametersMatchedWithArgs[paramIndex] = true;
                         }
                     }
+
+                    // if ref is passed and it is not a ref-like, check ref escape, 
+                    // since it is the same or narrower than val escape and callee may wrap the ref into a val
+                    var argument = args[argIndex];
+                    var valid = effectiveRefKind != RefKind.None && argument.Type.IsByRefLikeType == false ?
+                                        CheckRefEscape(argument.Syntax, argument, scopeOfTheContainingExpression, escapeTo, false, diagnostics) :
+                                        CheckValEscape(argument.Syntax, argument, scopeOfTheContainingExpression, escapeTo, false, diagnostics);
+
+                    if (!valid)
+                    {
+                        var paramIndex = argsToParamsOpt.IsDefault ? argIndex : argsToParamsOpt[argIndex];
+                        var parameterName = parameters[paramIndex].Name;
+                        Error(diagnostics, ErrorCode.ERR_CallArgMixing, syntax, symbol, parameterName);
+                        return false;
```

Comment:
Consider using `finally` block to free `inParametersMatchedWithArgs` in each of these methods.